Validate type consistency in reintepret cast sizes

yaniv217 opened the pull request:
Ensure that when peforming a reinterpret cast, the expected size and the result size are of the same type. Emit an error if one of the dimensions has a static size and the corresponding dimension has a dynamic size in the other.

yaniv217 committed:
Validate type consistency in reintepret cast sizes

Ensure that when peforming a reinterpret cast, the expected size and
the result size are of the same type. Emit an error if one of the
dimensions has a static size and the corresponding dimension has a
dynamic size in the other.

matthias-springer commented on mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp:
```diff
@@ -1835,6 +1835,15 @@ LogicalResult ReinterpretCastOp::verify() {
+    if (ShapedType::isDynamic(resultSize) &&
+        !ShapedType::isDynamic(expectedSize))
+      return emitError(
+          "expectedSize is static but received a dynamic resultSize ");
```
nit: It's a bit difficult to understand what "expectedSize" means in the error message. Maybe: `specified size is static, but result type dimension is dynamic`

yaniv217 replied:
I agree that it sounds more understandable but it seems that in the context of these functions they consistently use expected size and not specified size to name.

matthias-springer commented on mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp:
```diff
@@ -1835,6 +1835,15 @@ LogicalResult ReinterpretCastOp::verify() {
+    if (!ShapedType::isDynamic(resultSize) &&
+        ShapedType::isDynamic(expectedSize))
+      return emitError(
+          "expectedSize is dynamic but received a static resultSize ");
```
same

matthias-springer commented on mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp:
```diff
@@ -1835,6 +1835,15 @@ LogicalResult ReinterpretCastOp::verify() {
   // Match sizes in result memref type and in static_sizes attribute.
   for (auto [idx, resultSize, expectedSize] :
        llvm::enumerate(resultType.getShape(), getStaticSizes())) {
+    // Check that dynamic sizes are not mixed with static sizes
```
Can you add the same check for the offset and the strides?

yaniv217 replied:
Yes!

matthias-springer reviewed (commented):
Nice!

yaniv217 committed:
Validate type consistency in reintepret cast offsets

Ensure that when peforming a reinterpret cast, the expected offset and
the result offset are of the same type. Emit an error if one of the
dimensions has a static offset and the corresponding dimension has a
dynamic offset in the other. Delete previous test that is a specific
instance of this case.

yaniv217 committed:
Merge branch 'main' into fix/validate-reinterpret-cast-sizes
